realm/realm-library/src/objectServer/java/io/realm/PermissionManager.java

```diff
@@ -40,34 +42,37 @@
 public class PermissionManager implements Closeable {
 
     // Reference counted cache equivalent to how Realm instances work.
-    private static ThreadLocal<PermissionManager> permissionManager = new ThreadLocal<PermissionManager>() {
-        @Override
-        protected PermissionManager initialValue() {
-            return null;
-        }
-    };
+    private static Map<String, ThreadLocal<Cache>> cache = new HashMap<>();
 
-    private static ThreadLocal<Integer> permissionManagerInstanceCounter = new ThreadLocal<Integer>() {
-        @Override
-        protected Integer initialValue() {
-            return 0;
-        }
-    };
+    private static class Cache {
+        public PermissionManager pm = null;
+        public Integer instanceCounter = Integer.valueOf(0);
+    }
+
+    private static final Object cacheLock = new Object();
 
     /**
      * Return a thread confined, reference counted instance of the PermissionManager.
      *
      * @param syncUser user to create the PermissionManager for.
      * @return a thread confined PermissionManager instance for the provided user.
      */
-    static synchronized PermissionManager getInstance(SyncUser syncUser) {
-        PermissionManager pm = permissionManager.get();
-        if (pm == null) {
-            pm = new PermissionManager(syncUser);
-            permissionManager.set(pm);
+    static PermissionManager getInstance(SyncUser syncUser) {
+        synchronized (cacheLock) {
+            String userId = syncUser.getIdentity();
+            ThreadLocal<Cache> threadLocalCache = cache.get(userId);
+            if (threadLocalCache == null) {
+                threadLocalCache = new ThreadLocal<>();
+                threadLocalCache.set(new Cache());
+                cache.put(userId, threadLocalCache);
+            }
+            Cache c = threadLocalCache.get();
```

Comment:
In old code, you checked only `threadLocalCache` but also `c` need to be checked since you allocated `Cache` object only for the first thread.

But, current code is fine! `initialValue()` is executed on every thread.